src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs

```diff
@@ -2689,6 +2689,270 @@ var projectUniqueNamesForBuildIntToUpdate
+
+                // Create a copy to avoid modifying the original spec which may be shared.
+                var updatedPackageSpec = originalPackageSpec.Clone();
+
+                // If the lock file does not exist, restore before starting the operations
+                if (originalLockFile == null)
+                {
+                    var originalRestoreResult = await DependencyGraphRestoreUtility.PreviewRestoreAsync(
+                        SolutionManager,
+                        buildIntegratedProject,
+                        originalPackageSpec,
+                        dependencyGraphContext,
+                        providerCache,
+                        cacheModifier,
+                        sources,
+                        nuGetProjectContext.OperationId,
+                        logger,
+                        token);
+
+                    originalLockFile = originalRestoreResult.Result.LockFile;
+                }
+
+                lockFileLookup[buildIntegratedProject.MSBuildProjectPath] = originalLockFile;
+
+                foreach (var action in nuGetProjectActions)
+                {
+                    if (action.NuGetProjectActionType == NuGetProjectActionType.Uninstall)
+                    {
+                        // Remove the package from all frameworks and dependencies section.
+                        PackageSpecOperations.RemoveDependency(updatedPackageSpec, action.PackageIdentity.Id);
+                    }
+                    else if (action.NuGetProjectActionType == NuGetProjectActionType.Install)
+                    {
+                        if (updatedPackageSpec.RestoreMetadata.ProjectStyle == ProjectStyle.PackageReference)
+                        {
+                            PackageSpecOperations.AddOrUpdateDependency(updatedPackageSpec, action.PackageIdentity, updatedPackageSpec.TargetFrameworks.Select(e => e.FrameworkName));
+                        }
+                        else
+                        {
+                            PackageSpecOperations.AddOrUpdateDependency(updatedPackageSpec, action.PackageIdentity);
+                        }
+                    }
+
+                    updatedNugetPackageSpecLookup[buildIntegratedProject.MSBuildProjectPath] = updatedPackageSpec;
+                    dependencyGraphContext.PackageSpecCache[buildIntegratedProject.MSBuildProjectPath] = updatedPackageSpec;
+                }
+            }
+
+            // Restore based on the modified package specs for projects. This operation does not write the lock files to disk.
+            // This operation restores in parallel.
+            var restoreResults = await DependencyGraphRestoreUtility.PreviewRestoreProjectsAsync(
+                SolutionManager,
+                buildIntegratedProjects,
+                updatedNugetPackageSpecLookup.Values,
+                dependencyGraphContext,
+                providerCache,
+                cacheModifier,
+                allSources,
```

Comment:
Since `allSources`, as the name suggests, is all sources from all projects, including those that are defined to be specific to one project, I'm concerned that this means it is technically incorrect. Imagine a solution with two projects, P1 and P2. P2 has an additional source defined in its project file, but P1 does not.  This preview restore will use P2's additional source with P1, which might change the resolved graph in a way that's impossible when restoring P1 by itself (say, on the command line).

Reply:
@nkolev92 Can you comment on this? Now we're passing `allSources` for all projects same time to RestoreRunner.

Reply:
It looks `allsources` used only for creating requests [L227-L239](https://github.com/NuGet/NuGet.Client/blob/5e9b137d616dfcdbe4d4cfb537b23dfa592040a3/src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs#L227-L239) for RestoreRunner, but once it's passed to actual [RestoreRunner](https://github.com/NuGet/NuGet.Client/blob/5e9b137d616dfcdbe4d4cfb537b23dfa592040a3/src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs#L240)  it doesn't care about this anymore. It still reads source information from each project itself.

Reply:
What does "should be good" mean? Even though sources specific to project A are passed to project B, project B will ignore it and only use the sources that are really configured for project B?

In that case we should change this method to not all all sources from all projects, to reduce the confusion. Maybe with a one line comment explaining that project-specific sources will still be resolved in RestoreRunner.

Reply:
Added 1 line comment for DependencyGraphRestoreUtility.PreviewRestoreProjectsAsync#[Line192](https://github.com/NuGet/NuGet.Client/blob/68b91402c84fcda73733c9823b7bf97c86be73ce/src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs#L192).

Reply:
> Even though sources specific to project A are passed to project B, project B will ignore it and only use the sources that are really configured for project B?

In PR, it's for re-use of the SourceRepository object.

 > In that case we should change this method to not all all sources from all projects, to reduce the confusion. Maybe with a one line comment explaining that project-specific sources will still be resolved in RestoreRunner.

Yep that's what I mean by the method xml clarifying it :D

My bad, I realized how lacking my comment was.